pages/login/index.js

```diff
@@ -2,68 +2,74 @@ import Head from 'next/head';
+import React, { useContext, useState } from 'react';
 
 import Layout from '../../components/Layout';
 
 import { useAuth } from '../../contexts/auth';
-import { useWallet } from '../../hooks/useWallet';
-
+import { WalletDispatch, WalletState } from '../../contexts/wallet';
 import api from '../../utils/api';
 import { __ } from '../../utils/helpers';
 
 const Login = () => {
-  const { account, signMessage, isWalletConnected, connectWallet } = useWallet()
+  const { injected } = useContext(WalletState);
+  const { signMessage } = useContext(WalletDispatch);
 
   const router = useRouter();
-  const { isAuthenticated, login, setAuthentification, setError } = useAuth();
+  const { isAuthenticated, login, setAuthentification } = useAuth();
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
-  const [ shouldFollowUpOnConnectAndSign, setShouldFollowUpOnConnectAndSign ] = useState('')
+  const [error, setError] = useState('');
 
-  if (isAuthenticated && typeof window.location !== 'undefined') {
+  if (isAuthenticated) {
     router.push('/');
-    // For some reason, cache needs to get reset.
-    // window.location.href = decodeURIComponent(router.query.back || '/');
   }
 
-
-  const executeRestOfSignInWithWallet = async() => {
+  const signInWithWallet = async (walletAddress) => {
     try {
-      const { data: { nonce } } = await api.post('/auth/web3/pre-sign', { walletAddress: account });
-      const message =  `Signing in with code ${nonce}`;
-      const signedMessage = await signMessage(message)
       const {
-        data: { access_token: token, results: user },
-      } = await api.post('/auth/web3/login', {
-        signedMessage,
-        walletAddress: account,
-        message
-      });
-      setAuthentification(user, token)
-    } catch (error) {
-      setError(error.message);
+        data: { nonce },
+      } = await api.post('/auth/web3/pre-sign', { walletAddress });
+      const message = `Signing in with code ${nonce}`;
+      const signedMessage = await signMessage(message, walletAddress);
+      if (signedMessage) {
+        const {
+          data: { access_token: token, results: user },
+        } = await api.post('/auth/web3/login', {
+          signedMessage,
+          walletAddress,
+          message,
+        });
+        if (error) {
+          setError(error);
+          return;
+        }
+        setAuthentification(user, token);
+      }
+    } catch (e) {
+      if (e.response?.status === 401) {
+        setError(e.response.data.error);
+        return;
+      }
+      console.error(e);
     }
-  }
+  };
 
-  const walletConnectAndSignInFlow = async () => {
-    setShouldFollowUpOnConnectAndSign(true)
-    if(!isWalletConnected){
-      connectWallet()
-    }else{
-      executeRestOfSignInWithWallet()
+  const walletConnectAndSignInFlow = async (event) => {
+    event.preventDefault();
+    const activated = await injected.activate();
```

Comment:
@juliencharrel this is what `activate` from `useWeb3React` is missing - returning the address that has been activated. However, injected.activate() is doing the job quite well